### ISSUE TITLE
How to preserve trace context on http redirect ?

### ISSUE THREAD

**Issue by dmitriyse** [w3c/trace-context#438]
According to the internet knowledge: https://stackoverflow.com/questions/1969709/how-to-forward-headers-on-http-redirect
HTTP clients will not transfer headers on a redirect. 
But the redirect case also needs to preserve the trace context.
We have the next options:
1) Improve clients(not only browsers) to preserve the trace context headers.
2) Use alternative storage for the trace context information (query string, cookies, maybe something else)

In a long term, both options should be standardized. 

P.S. In a real-world app, I currently need to invent my own support through query string.

**Comment by SergeyKanzhelev** [comment]
> P.S. In a real-world app, I currently need to invent my own support through query string.

In this app, do you intent to send a new `parent-id` with each request (one resulting in 302 and one resulting in actual request) or want to reuse the same one? Both scenarios are quite valid depending on the semantics of these two calls. I *think* different `parent-id` would be more common ask. This way issues with 302 logic can be troubleshooted. If this is a true assumption, there is no special standardization is required. Curious if the same `parent-id` is something you need and why.

Or perhaps you seek for other standardization that will correlate 302 and 200 requests. Please comment. Thank you!

**Comment by dmitriyse** [comment]
I might not understand correctly. But we need to treat the 302 response -> Client -> "redirect target" request as a logical call to a different node. So the parent-id should be regenerated by the initial 302 response.

I haven't implemented currently the universal application-wide solution for the redirects tracing. Just added TraceID parameter to the most important redirect targets.

**Comment by dmitriyse** [comment]
The problem is to agree between two different software from different vendors on standard argument names:
ParentID = _tc_pid  
TraceID = _tc_tid
Example: 
```myapp1.com/myredirect``` ---->
```myapp2.com/mytarget?_tc_pid=<ParentID>&_tc_tid=<TraceID>```

The standardization will help to make different applications compatible out of the box.

**Comment by SergeyKanzhelev** [comment]
> Example:
> `myapp1.com/myredirect` ---->
> `myapp2.com/mytarget?_tc_pid=<ParentID>&_tc_tid=<TraceID>`

hm, do you expect the response from `myapp1.com/myredirect` to be an input to the second call? If not, there are no query string parameters needed, second call may use the same http headers protocol.

**Comment by dmitriyse** [comment]
If you have w3c headers to call
```myapp1.com/myredirect```
then
```myapp2.com/mytarget```
will have a chance to receive those headers.

But in a case where
```External Client-Side software``` ---call-no-w3c-headers--> ```myapp1.com/myredirect```
We need some way how to track the call chain between
```myapp1.com/myredirect``` and ```myapp2.com/mytarget```

In other words ```myapp2.com/mytarget``` should understand that it was redirected from ```myapp1.com/myredirect``` and know the context of the ```myapp1.com/redirect``` processing

To achieve this when doing redirect ```myapp1.com/myredirect``` should initiate w3c trace-context itself and somehow pass the identifiers to ```myapp2.com/mytarget```

**Comment by dyladan** [comment]
If the client understands W3C trace context this is not needed. The client can easily ensure these calls are both in the same trace.

- `myclient` starts a trace 1
- `myclient` makes a request to `myapp1.com/myredirect` with `traceid=1, spanid=1`
- `myapp1.com/myredirect` returns `302 myapp2.com/mytarget`
- `myclient` makes a request to `myapp2.com/mytarget` with `traceid=1, spanid=2`

If your goal is to be able to show that `myapp2.com/mytarget` was caused by `myapp1.com/myredirect`, you can easily do this by attaching metadata to the spans either from `myclient`, `myapp1.com/myredirect`, or both. I would argue, however, that these spans do not have a parent/child relationship anyways because `myclient` initiated both calls.

In other words, you are asking for:

```
myclient
	|
	*-> myapp1.com/myredirect
		|
		*-> myapp2.com/mytarget
```

but I think the more appropriate graph is:

```
myclient
	|
	*-> myapp1.com/myredirect
	|
	*-> myapp2.com/mytarget
```

**Comment by dmitriyse** [comment]
@dyladan , I agree with you. If you can control ```myclient``` then there is no problem, once it aware about w3c tracing and provides traceid.
But I am describing the case where you cannot control ```external-client``` which don't care about w3c tracing.

```
external-client    (unfortunately not myclient)
             | 
             | <-- no tracing context
             | 
             *myapp1.com/myredirect    
                      |
                      | <-- tracing problem
                      |
                      *->myapp2.com/mytarget

**Comment by yurishkuro** [comment]
If client does not support tracing, there's no way to correlate /redirect and /target calls, because the response from /redirect goes back to the client, which will drop the ball and send a new request that will look like a new trace.

**Comment by dmitriyse** [comment]
```myapp1.com/myredirect``` can control the redirect target and inject additional key=value pairs in the redirect target query string.
Example ```myapp1.com/redirect``` --->302--> ```myapp2.com/mytarget?target_data1=abc&w3c_traceid=12323&w3c_parent=12323``` assume ```myapp2.com/mytarget``` ignores unknown query arguments, it will accept only ```target_data1```, but ```w3c_traceid``` and ```w3c_parent``` will be captured by the w3c tracing middleware.
P.S.   ```w3c_traceid``` and ```w3c_parrent``` are not the proposed names for this purpose, I used these keys only to express the idea.

**Comment by yurishkuro** [comment]
Are you suggesting that TraceContext spec (which is for HTTP headers) also gives a recommendation to return trace context as part of the redirect URL?

If we did go in that direction, the actual field names would be the same, traceparent and tracestate.

**Comment by danielkhan** [comment]
I'm closing this. I think a general sentiment (and mine) is that the redirecting entity needs to propagate trace context as a redirect is technically not different from any other hop in a system.

**Comment by Mart-Bogdan** [comment]
Hello, all, I'm sorry for reopening old discussion.

@danielkhan wrote:

> I'm closing this. I think a general sentiment (and mine) is that the redirecting entity needs to propagate trace context as a redirect is technically not different from any other hop in a system.

The problem is, that browsers do not care about `traceparent` they don't geneerate it, and won't preserve it (and we can't actually trust it a lot in case if it would be provided).

Would be nice if we can indeed add some our field to query parameters upon redirect. Useful for example if redirect is required to log-in user. 

I guess It can be unstandartized, but in this case it won't be implemented by major web frameworks. 

For example aspnet as of 8.0 allows to use only headers in user provided implementation of `DistributedContextPropagator` so no way to access query string. 

My point is, this is quite useful feature to have. I don't know if this spec is the place for such feature. And indeed having same name in query parameters might be problematic.
